### ISSUE TITLE
[API Proposal]: GetFunctionPointer for delegates

### ISSUE THREAD

**Issue by IllidanS4** [dotnet/runtime#78846]
### Background and motivation

This is a broader counterpart to #45232 for delegates. As far as I know, there is no supported way to get the managed function pointer from a delegate in a fashion that would make it callable through `delegate*` in C# (for static methods, you can obtain it via reflection from `_methodPtrAux` or `interp_method` on Mono).

I suggest a method like `System.Runtime.CompilerServices.RuntimeHelpers.GetFunctionPointer(Delegate)`, returning `IntPtr` or perhaps `void*` to make it more "unsafe". For delegates without a target, it should be equivalent to obtaining the inner pointer.

For delegates with a target, there are three options here:

* Obtain "`_methodPtr`", ignoring the target and assuming that the user provides it when calling the delegate.
* Throw an exception, since the first option would mean that the result depends on something usually irrelevant when calling a delegate. The user would have to re-bind the method without a target first.
* Generate a "thunk" in a fashion similar to `Marshal.GetFunctionPointerForDelegate`, wrapping the target and the function pointer. I am most in favour of this option, since it is consistent with calling the actual delegate normally, even though it is non-trivial to implement. This would give it a considerable advantage over `RuntimeMethodHandle.GetFunctionPointer` (other than this working for dynamic methods).

The resulting pointer should be valid at least for as long as the delegate exists (comparable to the lifetime of an interior pointer to an object). This would come in handy as a general method to efficiently handle all situations and improve [interoperability between function pointers and delegates](https://github.com/dotnet/csharplang/discussions/3680) without having to deal with all the peculiarities.

### API Proposal

```csharp
namespace System.Runtime.CompilerServices;

public static class RuntimeHelpers
{
    public static void* GetFunctionPointer(Delegate function);
}
```

### API Usage

```csharp
Action<string> func = Console.WriteLine;
var ptr = (delegate*<string, void>)System.Runtime.CompilerServices.RuntimeHelpers.GetFunctionPointer(func);
ptr("Hello world!");
```

### Alternative Designs

Another option is to make `RuntimeMethodHandle.GetFunctionPointer` work for dynamic methods, which would at least fix the existing issues, but being able to obtain the pointer for *any* delegate would still be useful (considering that's something generally impossible to do in languages like C or C++).

### Risks

There are potential issues with the lifetime of the pointer which are similar to `Marshal.GetFunctionPointerForDelegate`, however I think that since the user has to use unsafe context in the first place anyway, this is something to be expected.

**Comment by DaZombieKiller** [comment]
The `MethodHandle` property is currently disallowed on `DynamicMethod` because it supposedly cannot be tracked by the GC, but I'm skeptical of this still being the case. The current CoreCLR implementation of `RuntimeMethodHandle` just stores an `IRuntimeMethodInfo` instance, which should be capable of keeping the `DynamicMethod` instance alive.

Once the `RuntimeMethodHandle` is converted to an `IntPtr` or a function pointer, it can no longer be tracked, but that is no different to the behavior you get with any other kind of method handle.

Maybe someone from the .NET team can clear up whether or not this is still a problem for `DynamicMethod`?

Note that for pointers to instance methods, there is currently no safe way to directly invoke such a function pointer from C# because we cannot declare managed function pointers with the `HasThis` calling convention. So even if you could use `GetFunctionPointer`, that doesn't mean you would be able to actually call it without additional work.

The theoretical syntax to allow that would be along the lines of:
```cs
delegate* managed[HasThis]<object, int> fp = &obj.GetHashCode;
```

On CoreCLR there is an internal [GetMethodDescriptor](https://github.com/dotnet/runtime/blob/4832236b4820ca83bbc9ce96003c2029b85c6ce2/src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.CoreCLR.cs#L82-L104) method that you can invoke through reflection to get a `RuntimeMethodHandle` for a `DynamicMethod`, but that is of course entering unsupported territory.

**Comment by IllidanS4** [comment]
> Note that for pointers to instance methods, there is currently no safe way to directly invoke such a function pointer from C# because we cannot declare managed function pointers with the HasThis calling convention. So even if you could use GetFunctionPointer, that doesn't mean you would be able to actually call it without additional work.

Yeah, that would be more complicated, hence I was thinking more about using a thunk that removes `HasThis` in such a case. It seems `_methodPtrAux` actually already contains such a thunk, so it's just a matter of exposing it like I suggested.

**Comment by lambdageek** [comment]
One additional consideration for this API proposal is that it may not be very AOT friendly.  Specifically, if we have to generate a thunk (ie the third option for methods with a target), it's unclear that we will know at AOT time that one is required.  Mono previously solved this by adding a `[MonoPInvokeCallback(typeof(Action<int>))]` attribute on methods - specifying the delegate that the method would be wrapped in and giving a hint to the AOT compiler that a thunk will be needed.

I think even options 1 and 2 may need some runtime code gen on some calling conventions to adopt between the caller (expected to pass a target object) and the callee (expecting a `this` pointer), which is also not AOT friendly.
